File: notebookspython/18-add_without_arithmetic.py
class Solution:
    def add(self, a: int, b: int) -> int:
        """Add two 32-bit signed integers using bit operations only."""
        mask = 0xFFFFFFFF
        max_int = 0x7FFFFFFF

        while b != 0:
            carry = (a & b) << 1
            a = (a ^ b) & mask
            b = carry & mask

        if a <= max_int:
            return a
        return ~(a ^ mask)


if __name__ == "__main__":
    sol = Solution()

    # Case 1: two positive numbers
    assert sol.add(1, 2) == 3

    # Case 2: zero keeps the other value unchanged
    assert sol.add(5, 0) == 5

    # Case 3: opposite signs cancel out
    assert sol.add(-1, 1) == 0

    # Case 4: two negative numbers
    assert sol.add(-3, -4) == -7

    # Case 5: larger positive numbers
    assert sol.add(123, 456) == 579

    # Case 6: follows 32-bit signed integer overflow behavior
    assert sol.add(0x7FFFFFFF, 1) == -0x80000000

    print("All test cases passed ✅")

File: notebooksjava/_18AddWithoutArithmetic.java
public class _18AddWithoutArithmetic {

    public int add(int a, int b) {
        // 用异或合并无进位结果，用与和左移计算进位。
        while (b != 0) {
            int carry = (a & b) << 1;
            a = a ^ b;
            b = carry;
        }

        return a;
    }

    public static void main(String[] args) {
        _18AddWithoutArithmetic sol = new _18AddWithoutArithmetic();

        // Case 1: 两个正数
        assert sol.add(1, 2) == 3 : "case 1 failed";

        // Case 2: 0 不改变另一个值
        assert sol.add(5, 0) == 5 : "case 2 failed";

        // Case 3: 符号相反时相互抵消
        assert sol.add(-1, 1) == 0 : "case 3 failed";

        // Case 4: 两个负数
        assert sol.add(-3, -4) == -7 : "case 4 failed";

        // Case 5: 较大的正数
        assert sol.add(123, 456) == 579 : "case 5 failed";

        // Case 6: 遵循 Java int 溢出行为
        assert sol.add(0x7fffffff, 1) == 0x80000000 : "case 6 failed";

        System.out.println("All test cases passed ✅");
    }
}
